schema_editor/app/scripts/schemas/schemas-service.js

```diff
@@ -45,21 +61,109 @@
          */
         function fieldFromKey(fieldKey, fieldOptions) {
             fieldOptions = fieldOptions || {};
-            var field = null;
-            switch (fieldKey) {
-                case 'text':
-                    field = textField(fieldOptions);
-                    break;
-                case 'selectlist':
-                    field = selectList(fieldOptions);
-                    break;
-                case 'image':
-                    field = imageUploader(fieldOptions);
-                    break;
-                default:
-                    throw 'key must be one of Schemas.FieldTypes';
+            if (module.FieldTypes[fieldKey] === undefined) {
+                throw 'key must be one of Schemas.FieldTypes';
             }
-            return field;
+            return module.FieldTypes[fieldKey].create(fieldOptions);
+        }
+
+        /**
+         * Encode a string so that it complies with the JSON Pointer spec
+         * http://tools.ietf.org/html/draft-pbryan-zyp-json-pointer-02
+         * This is what is used in $ref to specify sub-schema paths
+         * @param {string} str
+         * @return {string} Currently just returns encodeURIComponent(string)
+         */
+        function encodeJSONPointer(str) {
+            // TODO: Technically, we should be doing this:
+            // return encodeURIComponent(str);
+            // But json-editor doesn't seem to support that properly.
+            return str;
+        }
+
+        /**
+         * Converts part of the output of a Schema Entry Form (data about fields)
+         * into a snippet designed to be inserted into the 'properties' key of a Data Form Schema
+         * which would be used for data entry / editing. Does not perform validation.
+         * @param {object} fieldData The value of one Schema Entry Form field, specifying required,
+         *      searchable, possible values, etc.
+         * @return {object} A snippet designed to be inserted into the 'properties' of a Data Form
+         *      Schema
+         */
+        // TODO: This monolithic function isn't great; investigate more modular options
+        function _propertyFromSchemaFieldData(fieldData) {
+            var propertyDefinition = {};
+            propertyDefinition.type = module.FieldTypes[fieldData.fieldType].jsonType;
+
+            if (fieldData.fieldType === 'selectlist') {
+                propertyDefinition.enum = fieldData.fieldOptions;
+            }
+            if (fieldData.fieldType === 'image') {
```

Comment:
I suppose this can be an 'else if'. Though I agree that something more modular would be great -- I'll give that some thought, but I bet we can whip something up using prototypical inheritance.